### PR TITLE
CMF-2: Create DB structure

### PR DESCRIPTION
**Pull request by CheAlex**
http://jira.grossum.com/browse/CMF-2

### CODE REVIEW

File: Entity/Repository/StaticPageRepository.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+class StaticPageRepository extends EntityRepository
+{
+    /**
+     * @return StaticPage[]
+     */
+    public function findAllEnabled()
+    {
+        return $this
+            ->_em
```
Comment (malavsky) [inline Entity/Repository/StaticPageRepository.php]: Please, remove `->_em->getRepository('GrossumStaticPageBundle:StaticPage')`

File: Resources/config/validation.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+Grossum\StaticPageBundle\Entity\StaticPage:
+    properties:
+        title:
+            - NotBlank: ~
+#        slug:
```
Comment (malavsky) [inline Resources/config/validation.yml]: if unuse, delete it

File: Entity/Repository/StaticPageRepository.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace Grossum\StaticPageBundle\Entity\Repository;
+
+use Doctrine\ORM\EntityRepository;
+use Grossum\StaticPageBundle\Entity\StaticPage;
```
Comment (malavsky) [inline Entity/Repository/StaticPageRepository.php]: separate uses please

File: Entity/EntityManager/StaticPageManager.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+namespace Grossum\StaticPageBundle\Entity\EntityManager;
+
+use Doctrine\Common\Persistence\ObjectManager;
+use Grossum\CoreBundle\Entity\EntityTrait\SaveUpdateInManagerTrait;
```
Comment (malavsky) [inline Entity/EntityManager/StaticPageManager.php:7]: separate uses, please

File: Entity/EntityManager/StaticPageManager.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        $this->objectManager = $objectManager;
+    }
+
+    /**
+     * @return StaticPageRepository
+     */
+    private function getRepository()
+    {
+        return $this->objectManager->getRepository('GrossumStaticPageBundle:StaticPage');
```
Comment (malavsky) [inline Entity/EntityManager/StaticPageManager.php]: Please, 

```
if (!$this->repository) {
    $this->repository = $this->objectManager->getRepository($this->staticPageClass);
}

return $this->repository;
```

File: Admin/StaticPageAdmin.php [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+class StaticPageAdmin extends Admin
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function configureFormFields(FormMapper $formMapper)
+    {
+        $formMapper
+            ->add('enabled')
```
Comment (malavsky) [inline Admin/StaticPageAdmin.php]: Please, add title to field. Ithink it should be for all fields.
Please try to add title from translation file.

File: Admin/StaticPageAdmin.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            ->add('title', TextType::class)
```
Comment (malavsky) [inline Admin/StaticPageAdmin.php]: remove type for field I think it will be choose automaticly

File: Admin/StaticPageAdmin.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            ->add('parent')
+            ->add(
+                'slug',
+                TextType::class,
```
Comment (malavsky) [inline Admin/StaticPageAdmin.php]: remove type for field I think it will be choose automaticly

File: Admin/StaticPageAdmin.php [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+                ]
+            )
+            ->add(
+                'body',
+                CKEditorType::class,
+                [
+                    'required' => false
+                ]
+            )
+            ->add('position', IntegerType::class);
```
Comment (malavsky) [inline Admin/StaticPageAdmin.php]: remove type for field I think it will be choose automaticly

File: Entity/StaticPage.php [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+
+    /**
+     * @var StaticPage
+     */
+    private $parent;
+
+    public function __construct()
+    {
+        $this->children = new ArrayCollection();
+
```
Comment (malavsky) [inline Entity/StaticPage.php]: remove extra new line

File: Admin/StaticPageAdmin.php (hunk 6 above)
Comment (WhiteEagle88) [inline Admin/StaticPageAdmin.php]: @malavsky wrong, he's mean add label

File: Resources/config/doctrine/StaticPage.orm.yml [hunk 11]
```diff
@@ -0,0 +1,52 @@
+Grossum\StaticPageBundle\Entity\StaticPage:
+    type: entity
```
Comment (WhiteEagle88) [inline Resources/config/doctrine/StaticPage.orm.yml]: Please, change type on mappedSuperClass

File: Resources/config/doctrine/StaticPage.orm.yml [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    fields:
+        title:
+            type: string
+            length: 255
+        slug:
+            type: string
+            length: 255
+            gedmo:
+                slug:
+                    separator: _
```
Comment (WhiteEagle88) [inline Resources/config/doctrine/StaticPage.orm.yml]: change separator on hyphen

File: Resources/config/doctrine/StaticPage.orm.yml (hunk 11 above)
Comment (CheAlex) [inline Resources/config/doctrine/StaticPage.orm.yml]: This class should be entity because it has associations and mappedSuperClass can't has it

Comment (aivus) [inline Resources/config/doctrine/StaticPage.orm.yml]: I think, we will do something like this:
https://github.com/sonata-project/SonataMediaBundle/blob/f0e6752e633cff5598ec04925bfa51e1a60fe7f4/DependencyInjection/SonataMediaExtension.php#L200
https://github.com/GrossumUA/SonataEasyExtendsBundle/blob/master/DependencyInjection/Compiler/AddMapperInformationCompilerPass.php#L39
https://github.com/GrossumUA/SonataEasyExtendsBundle/blob/master/Mapper/DoctrineORMMapper.php

File: DependencyInjection/GrossumStaticPageExtension.php [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+        if (!isset($registeredBundles['GrossumCoreBundle'])) {
+            throw new LogicException('GrossumStaticPageBundle required GrossumCoreBundle');
+        } elseif (!isset($registeredBundles['IvoryCKEditorBundle'])) {
+            throw new LogicException('GrossumStaticPageBundle required IvoryCKEditorBundle');
+        }
+
+        $container->setParameter('grossum_static_page.entity.static_page.class', $config['class']['static_page']);
+
+        $loader = new Loader\YamlFileLoader($container, new FileLocator(__DIR__.'/../Resources/config'));
```
Comment (aivus) [inline DependencyInjection/GrossumStaticPageExtension.php]: Missed spaces around `.`

File: Entity/BaseStaticPage.php [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+
+    /**
+     * @var \DateTime
+     */
+    protected $updatedAt;
+
+    /**
+     * @var Collection|BaseStaticPage[]
+     */
+    protected $children;
```
Comment (aivus) [inline Entity/BaseStaticPage.php]: We should use plural form when use collections

File: Entity/BaseStaticPage.php [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+    {
+        return $this->parent;
+    }
+
+    /**
+     * @return string
+     */
+    public function __toString()
+    {
+        return $this->getTitle() ?: "Static page";
```
Comment (aivus) [inline Entity/BaseStaticPage.php]: Please, don't use double quotes without necessary

File: README.md [hunk 16]
```diff
@@ -34,6 +34,8 @@ class AppKernel extends Kernel
             // ...
 
             new Stof\DoctrineExtensionsBundle\StofDoctrineExtensionsBundle(),
+            new CoopTilleuls\Bundle\CKEditorSonataMediaBundle\CoopTilleulsCKEditorSonataMediaBundle(),
```
Comment (aivus) [inline README.md]: Why not a https://github.com/egeloen/IvoryCKEditorBundle ?

File: Resources/translations/messages.en.yml [hunk 17]
```diff
@@ -0,0 +1,11 @@
+grossum_static_page:
+    admin:
+        static_page:
+            label       : 'Static page'
```
Comment (aivus) [inline Resources/translations/messages.en.yml]: If you want to align it, please, format as:

```
label:        'Static page'
plural_label: 'Static pages'
```

File: Entity/BaseStaticPage.php (hunk 15 above)
Comment (WhiteEagle88) [inline Entity/BaseStaticPage.php]: What do you mean?

Comment (aivus) [inline Entity/BaseStaticPage.php]: Use ' instead of "

Comment (WhiteEagle88) [inline Entity/BaseStaticPage.php]: @aivus, fixed

**Comment by aivus** [comment]
Looks like ok for me.
@malavsky @WhiteEagle88, can we merge it?

File: DependencyInjection/GrossumStaticPageExtension.php [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+     * {@inheritdoc}
+     */
+    public function load(array $configs, ContainerBuilder $container)
+    {
+        $configuration = new Configuration();
+        $config        = $this->processConfiguration($configuration, $configs);
+
+        $registeredBundles = $container->getParameter('kernel.bundles');
+
+        if (!isset($registeredBundles['GrossumCoreBundle'])) {
```
Comment (aivus) [inline DependencyInjection/GrossumStaticPageExtension.php]: I think, we can use array for required bundles.
Like a:

```
protected $requiredBundles = ['GrossumCoreBundle', 'IvoryCKEditorBundle'];
```

and here something like:

```
foreach ($this->requiredBundles as $bundle) {
    if (!isset($registeredBundles[$bundle]))) {
        throw new LogicException(sprintf('GrossumStaticPageBundle required %s', $bundle));
    }
}
```

**Pull request merged**